include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -157,32 +161,291 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
 		*/
 		struct SCompilerOptions
 		{
+			inline void setCommonData(const SCompilerOptions& opt)
+			{
+				(*this) = opt;
+			}
+
+			virtual IShader::E_CONTENT_TYPE getCodeContentType() const { return IShader::E_CONTENT_TYPE::ECT_UNKNOWN; };
+
 			IShader::E_SHADER_STAGE stage = IShader::E_SHADER_STAGE::ESS_UNKNOWN;
 			E_SPIRV_VERSION targetSpirvVersion = E_SPIRV_VERSION::ESV_1_6;
 			const ISPIRVOptimizer* spirvOptimizer = nullptr;
 			core::bitflag<E_DEBUG_INFO_FLAGS> debugInfoFlags = core::bitflag<E_DEBUG_INFO_FLAGS>(E_DEBUG_INFO_FLAGS::EDIF_SOURCE_BIT) | E_DEBUG_INFO_FLAGS::EDIF_TOOL_BIT;
 			SPreprocessorOptions preprocessorOptions = {};
+		};
+
+		class CCache final : public IReferenceCounted
+		{
+		public:
+			using hash_t = std::array<uint64_t, 4>;
+			static auto const CONTAINER_JSON_SIZE_BYTES = sizeof(uint64_t) / sizeof(uint8_t); // It's obviously 8
 
-			void setCommonData(const SCompilerOptions& opt)
+			struct SEntry
 			{
-				(*this) = opt;
+
+				struct SPreprocessingDependency
+				{
+					// Perf note: hashing while preprocessor lexing is likely to be slower than just hashing the whole array like this 
+					inline SPreprocessingDependency(const system::path& _requestingSourceDir, const std::string_view& _identifier, const std::string_view& _contents) :
+						requestingSourceDir(_requestingSourceDir), identifier(_identifier), contents(_contents)
+					{
+						assert(!_contents.empty());
+						const auto reqDirStr = requestingSourceDir.make_preferred().string();
+						std::vector<char> hashable;
+						hashable.insert(hashable.end(), reqDirStr.data()[0], reqDirStr.data()[reqDirStr.size()]);
+						hashable.insert(hashable.end(), identifier.data()[0], identifier.data()[identifier.size()]);
+						hashable.insert(hashable.end(), _contents.data()[0], _contents.data()[_contents.size()]);
```

Comment:
what on earth are you doing here?

I'm sure that insert takes position, then a begin and end iterator of the source, you're passing random out of bound chars. Some Rust-isms that don't translate to C++?

P.S. `hashable` should be created with the correct size of `reqDirStr.size()+identifier.size()+_contents.size()` so it does not need to reallocate